src/PluginClientBuilder.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ?PluginClient */
+    private $client;
+
+    public function addPlugin(Plugin $plugin): self
+    {
+        $builder = new self();
+
+        $builder->client = null;
+        $builder->plugins = $this->plugins;
+        $builder->plugins[get_class($plugin)] = $plugin;
```

Comment:
This limits the user to add only one instance of the plugin to the chain. Even worse, there is no warning that a plugin is overwritten.

Reply:
Originally I had a `string $name` argument, but in @Nyholm's pr on the builder, I saw he was using the fqcn instead.

I'll revert back to the plugin name instead.